### PR TITLE
Use Boost.Asio string_view if available.

### PR DESCRIPTION
**Pull request by Jiwan**
This PR tries to align **Boost.Beast**'s string_view with **Boost.Asio**. Since ``std::string_view`` and ``std::experimental::string_view`` do not have a ``clear`` member function, ``a_string_view = string_view();`` is used instead ([see n4288](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4288.html)).

### CODE REVIEW

**Comment by Jiwan** [comment]
Note that I am still unsure if using the ``detail`` folder of **Boost.Asio** is wise.

File: include/boost/beast/core/string.hpp [hunk 1]
```diff
@@ -10,21 +10,45 @@
 #ifndef BOOST_BEAST_STRING_HPP
 #define BOOST_BEAST_STRING_HPP
 
+#include <boost/asio/detail/string_view.hpp>
```
Comment (vinniefalco) [inline include/boost/beast/core/string.hpp]: Hmm... I don't know about this! Including a detail header is not best practices

File: include/boost/beast/core/string.hpp [hunk 2]
```diff
@@ -10,21 +10,45 @@
...
+
+#ifndef BOOST_ASIO_NO_STRING_VIEW
+# if defined(BOOST_ASIO_HAS_STD_STRING_VIEW)
+#  define BOOST_ASIO_HAS_STRING_VIEW 0
+# else
+#  define BOOST_ASIO_HAS_STRING_VIEW 1
```
Comment (vinniefalco) [inline include/boost/beast/core/string.hpp]: I don't think we are allowed to adjust these macros. The user needs to adjust them not our library.

File: include/boost/beast/core/string.hpp [hunk 3]
```diff
@@ -10,21 +10,45 @@
...
+# endif
+#endif
+
 #include <boost/beast/core/detail/config.hpp>
 #include <boost/version.hpp>
+
+#if BOOST_ASIO_NO_STRING_VIEW
 #include <boost/utility/string_view.hpp>
+#endif
```
Comment (vinniefalco) [inline include/boost/beast/core/string.hpp]: I think we need @chriskohlhoff to weigh in on this, to decide if he wants to support other libraries piggy-backing on the string view declaration.

File: include/boost/beast/core/string.hpp (hunk 2 above)
Comment (Jiwan) [reply include/boost/beast/core/string.hpp]: I was simplified things a bit more.

File: include/boost/beast/http/detail/rfc7230.hpp [hunk 4]
```diff
@@ -352,8 +352,8 @@ increment()
         {
             it = first;
         };
-    v.first.clear();
-    v.second.clear();
+    v.first = string_view();
+    v.second = string_view();
```
Comment (vinniefalco) [inline include/boost/beast/http/detail/rfc7230.hpp]: Consider this instead:
```
v.first = {};
v.second = {};
```

Comment (vinniefalco) [reply include/boost/beast/http/detail/rfc7230.hpp]: Actually, I'll make this change myself in its own commit.

Comment (Jiwan) [reply include/boost/beast/http/detail/rfc7230.hpp]: :+1:

**Comment by omartijn** [comment]
It'd be really great if this issue could be revisited. The fact that asio and beast use different string_view implementations feels sub-optimal. It's mostly small things, like not being able to directly use `boost::beast::http::message::operator[](boost::beast::http::field::host)` as parameter to `boost::asio::ip::tcp::resolver::async_resolve`.

It's trivial to work around, but it doesn't make the code more readable. In addition, I quite like the idea of using `std::string_view` (or `std::experimental::string_view`) when available.

**Comment by vinniefalco** [comment]
Yes I believe we can do something. In "stand-alone" (not implemented yet) mode Beast can use standalone Asio, require C++17, and use `std::string_view`.

Otherwise, Beast can require C++11 and use Boost.Asio. In this configuration, Beast will use `boost::string_view`, unless the Asio macro `BOOST_ASIO_HAS_STD_STRING_VIEW` is defined, in which case Beast will use `std::string_view` and require C++17.

**Pull request closed without merging**